regression/strings-smoke-tests/cprover/CProverString.java

```diff
@@ -0,0 +1,30 @@
+package org.cprover;
+
+public final class CProverString
+{
+    public static char charAt(String s, int i) {
+        if(0 <= i && i < s.length())
+            return s.charAt(i);
+        else
+            return '\u0000';
+    }
+
+    public static String substring(String s, int i) {
+        if(i <= 0)
+            return s;
+        else if(i >= s.length())
+            return "";
+        else
+            return s.substring(i);
```

Comment:
This I don't understand. Who handles that function if the preprocessing is disabled for `String.substring(I)`?

Reply:
the preprocessing is enabled for CProverString.substring instead, so this function is replace by the preprocessing

Reply:
So what's the purpose of this code? It won't be used in cbmc will it?

Reply:
I mean, you have to provide the functions, but why do you distinguish between cases?

Reply:
This is so that the programs that call CProverString.substring can be executed and the result of the execution will be the same as what jbmc expects it to be.
At the same time it documents what the internal substring exactly corresponds to.

Reply:
OK I see. It makes sense to allow these tests to still execute. The logic is the other way around compared to the models library, which makes it confusing. But I don't see an easy way around that.

Is there any reason why we return `s` when `i < 0`? The original implementation throws an exception, so `""` would probably make more sense?

Reply:
I was going to say that it's what the substring primitive of the solver does, but actually looking at the code it would generate a string of length `length - i` where the first characters can be anything...
I don't want to change the constraints in the solver in this PR because this would be out of scope, so what about appending `-i` times the '\u0000' character at the beginning?
I would fit in the specifications of the internal primitive (but the user should not rely on the first characters being '\u0000').

Reply:
Thanks for the info. Yes you can change it to have null chars, but it's not very important, as this code is only used when running the test files.